pkg/types/apisix/v1/types.go

```diff
@@ -468,17 +470,20 @@ func NewDefaultPluginConfig() *PluginConfig {
 
 // ComposeUpstreamName uses namespace, name, subset (optional) and port info to compose
 // the upstream name.
-func ComposeUpstreamName(namespace, name, subset string, port int32) string {
+func ComposeUpstreamName(namespace, name, subset string, port int32, resolveGranularity string) string {
 	pstr := strconv.Itoa(int(port))
 	// FIXME Use sync.Pool to reuse this buffer if the upstream
 	// name composing code path is hot.
 	var p []byte
-	if subset == "" {
-		p = make([]byte, 0, len(namespace)+len(name)+len(pstr)+2)
-	} else {
-		p = make([]byte, 0, len(namespace)+len(name)+len(subset)+len(pstr)+3)
+	plen := len(namespace) + len(name) + len(pstr) + 2
+	if subset != "" {
+		plen = plen + len(subset) + 1
+	}
+	if resolveGranularity != "" && resolveGranularity != types.ResolveGranularity.Endpoint {
```

Comment:
```suggestion
	if resolveGranularity != types.ResolveGranularity.Service {
```

Should we change this?

Reply:
I think it should be `resolveGranularity == types.ResolveGranularity.Service`, but should we support more ResolveGranularity?